cirq-google/cirq_google/engine/simulated_local_processor.py

```diff
@@ -0,0 +1,307 @@
+            del self._programs[program_id]
+
+    def get_calibration(self, calibration_timestamp_seconds: int) -> calibration.Calibration:
+        return self._calibrations[calibration_timestamp_seconds]
+
+    def get_latest_calibration(self, timestamp: int) -> Optional[calibration.Calibration]:
+        if not self._calibrations:
+            return None
+        return self._calibrations[max(self._calibrations)]
+
+    def get_current_calibration(self) -> Optional[calibration.Calibration]:
+        return self.get_latest_calibration(int(datetime.datetime.now().timestamp()))
+
+    def get_device(self, gate_sets: Optional[Iterable['Serializer']] = None) -> cirq.Device:
+        """Returns a `Device` created from the processor's device specification.
+
+        This method queries the processor to retrieve the device specification,
+        which is then use to create a `SerializableDevice` that will validate
+        that operations are supported and use the correct qubits.
+        """
+        return self._device
+
+    def get_device_specification(self) -> Optional[v2.device_pb2.DeviceSpecification]:
+        raise NotImplementedError
+
+    def health(self):
+        return 'OK'
+
+    def list_calibrations(
+        self,
+        earliest_timestamp: Optional[Union[datetime.datetime, datetime.date, int]] = None,
+        latest_timestamp: Optional[Union[datetime.datetime, datetime.date, int]] = None,
+    ) -> List[calibration.Calibration]:
+        earliest_timestamp_seconds = _date_to_timestamp(earliest_timestamp) or 0
+        latest_timestamp_seconds = (
+            _date_to_timestamp(latest_timestamp)
+            or datetime.datetime(datetime.MAXYEAR, 1, 1).timestamp()
+        )
+        return [
+            cal[1]
+            for cal in self._calibrations.items()
+            if earliest_timestamp_seconds <= cal[0] <= latest_timestamp_seconds
+        ]
+
+    def get_sampler(self, gate_set: Optional['Serializer'] = None) -> cirq.Sampler:
+        return self._sampler
+
+    def supported_languages(self) -> List[str]:
+        return VALID_LANGUAGES
+
+    def list_programs(
+        self,
+        created_before: Optional[Union[datetime.datetime, datetime.date]] = None,
+        created_after: Optional[Union[datetime.datetime, datetime.date]] = None,
+        has_labels: Optional[Dict[str, str]] = None,
+    ) -> List[AbstractLocalProgram]:
+        before_limit = created_before or datetime.datetime(datetime.MAXYEAR, 1, 1)
+        after_limit = created_after or datetime.datetime(datetime.MINYEAR, 1, 1)
+        labels = has_labels or {}
+        return list(
+            filter(
+                lambda program: after_limit < program.create_time() < before_limit
+                and all(
+                    (key in program.labels() and program.labels()[key] == labels[key])
+                    for key in labels
+                ),
+                self._programs.values(),
+            )
+        )
```

Comment:
Perhaps storing an intermediate list or two here to break up the big compound expression might be a little easier to parse as a reader ?

Reply:
Made an intermediate function instead.